packages/reactDom/src/ReactServerDOM.ml

```diff
@@ -96,93 +96,98 @@ module Model = struct
-                  let%lwt element = promise in
-                  context.pending <- context.pending - 1;
-                  context.push index (Chunk_value (to_payload element));
-                  if context.pending = 0 then context.close ();
-                  Lwt.return ());
-              `String (lazy_value index))
-      | Suspense { key; children; fallback } ->
-          (* TODO: Maybe we need to push suspense index and suspense node separately *)
-          let fallback = to_payload fallback in
-          suspense_node ~key ~fallback [ to_payload children ]
-      | Client_component { import_module; import_name; props; client = _ } ->
-          let id = use_chunk_id context in
-          let ref = component_ref ~module_:import_module ~name:import_name in
-          context.push id (Chunk_component_ref ref);
-          let client_props = client_props_to_json props in
-          node ~tag:(ref_value id) ~key:None ~props:client_props []
-      (* TODO: Dow we need to do anything with Provider and Consumer? *)
-      | Provider children -> to_payload children
-      | Consumer children -> to_payload children
-    and client_props_to_json props =
-      List.map
-        (fun (name, value) ->
-          match (name, (value : React.client_prop)) with
-          | name, Json json -> (name, json)
-          | name, Element element ->
-              (* TODO: Probably a silly question, but do I need to push this client_ref? (What if it's a client_ref?) In case of server, no need to do anything I guess *)
-              (name, to_payload element)
-          | name, Promise (promise, value_to_json) -> (
-              match Lwt.state promise with
-              | Return value ->
-                  let chunk_id = use_chunk_id context in
-                  let json = value_to_json value in
-                  (* TODO: Make sure why we need a chunk here *)
-                  context.push context.chunk_id (Chunk_value json);
-                  (name, `String (promise_value chunk_id))
-              | Sleep ->
-                  let chunk_id = use_chunk_id context in
-                  context.pending <- context.pending + 1;
-                  Lwt.async (fun () ->
-                      let%lwt value = promise in
-                      let json = value_to_json value in
-                      context.pending <- context.pending - 1;
-                      context.push chunk_id (Chunk_value json);
-                      if context.pending = 0 then context.close ();
-                      Lwt.return ());
-                  (name, `String (promise_value chunk_id))
-              | Fail exn ->
-                  (* TODO: Can we check if raise is good heres? *)
-                  raise exn))
-        props
-    in
-    let initial_chunk_id = get_chunk_id context in
-    context.push initial_chunk_id (Chunk_value (to_payload element));
-    if context.pending = 0 then context.close ()
+  let rec element_to_payload ~context element =
```

Comment:
@davesnx I think we cannot move the rec to an inner function because `client_values_to_json` is an `and` function that is used in `createActionResponse`